spring-cloud-gcp-autoconfigure/src/main/java/org/springframework/cloud/gcp/autoconfigure/datastore/GcpDatastoreAutoConfiguration.java

```diff
@@ -102,20 +105,27 @@
 
 	@Bean
 	@ConditionalOnMissingBean
-	public Datastore datastore() {
-		DatastoreOptions.Builder builder = DatastoreOptions.newBuilder()
-				.setProjectId(this.projectId)
-				.setHeaderProvider(new UserAgentHeaderProvider(this.getClass()))
-				.setCredentials(this.credentials);
-		if (this.namespace != null) {
-			builder.setNamespace(this.namespace);
-		}
-
-		if (this.host != null) {
-			builder.setHost(this.host);
-		}
+	public NamespaceProvider namespaceProvider() {
+		return () -> this.namespace;
+	}
 
-		return builder.build().getService();
+	@Bean
+	@ConditionalOnMissingBean(value = Datastore.class, parameterizedContainer = Supplier.class)
+	public Supplier<Datastore> datastore(Supplier<String> namespaceProvider) {
```

Comment:
I think you need to use `NamespaceProvider` here instead of `Supplier<String>` to avoid ambiguity with other `Supplier<String>` beans that might be in the application context.
In fact, is there anything wrong with using `NamespaceProvider` everywhere instead of `Supplier<String>`?

Reply:
Yes good catch. I made that new type and forgot to come back and replace it here.